### PR TITLE
esp8266: Add esp.espconn class 

### PR DESCRIPTION
**Pull request by atx**

### CODE REVIEW

**Comment by pfalcon** [comment]
Definitely thank you for working on this, the port exactly needed community contribution to move forward. But are you aware of https://github.com/micropython/micropython/issues/992 , did you follow it? It had some ideas and discussion on how to deal with esp8266 support, given that its API is not compliant with BSD sockets (on which Python's socket module is based).

>  * Due to the nature of the built-in scheduler, there is no blocking
>   support
> * Supports callbacks
> * As there is no way (that I know of) the espconn_regist_connectcb()
>   callback can recognize on which socket has the connection arrived,
>   only one listening function at a time is supported

Sorry, that's completely not how Python's socket (and MicroPython's usocket) modules work. That means that it cannot be "usocket" module. Actually "esp" module was introduced to capture such non-complient, esp8266-adhoc API: 6ec650b41fd340a1f5787a86ce3388b2d1ff5311

File: esp8266/modusocket.c [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+// tuple such as ('192.168.0.1', 8080).
+STATIC mp_obj_t usocket_format_inet_addr(uint8_t *ip, mp_uint_t port) {
+    mp_obj_t tuple[2] = {
+        tuple[0] = usocket_format_ipv4_addr(ip),
+        tuple[1] = mp_obj_new_int(port),
+    };
+    return mp_obj_new_tuple(2, tuple);
+}
+
+STATIC mp_obj_t call_function_1_protected(mp_obj_t fun, mp_obj_t arg) {
```
Comment (pfalcon) [inline esp8266/modusocket.c]: But there's already "call function protected": 5fb775a0c059d187366525ff66be1c892a7b3302 (and that's not the 1st case of it either). How many copies of that do we need?

**Comment by atalax** [comment]
Turns out I forgot to `%retab` in modesp.c, fixed
The constants are what `esp.status()` returns, they can be used to check in which state the Wi-Fi is in.
`call_function_N_protected` fns were moved to a new file.

Should I rename the socket class and move it under the esp module?

**Comment by pfalcon** [comment]
> call_function_N_protected fns were moved to a new file.

Thanks for quick turnaround, that's good start.

> Should I rename the socket class and move it under the esp module?

Yes, I'm definitely -1 on calling something "usocket" which doesn't adhere to common interface.

Generally, what's I'd suggest is to split these 2 commits into separate PR - they both would need more discussion, but why block one on the other?

Thanks.

**Comment by atalax** [comment]
> Generally, what's I'd suggest is to split these 2 commits into separate PR - they both would need more discussion, but why block one on the other?

Done, moved the usocket code to esp.espconn and separated the first commit to #1212. I also fixed random crashes due to missing `esp_espconn_connect_callback_client` and premature `gc_free`s in `.close()`.

**Comment by dpgeorge** [comment]
Thanks, really great to have some help with this port!

Looks good but I think there is some refactoring to do; will comment inline.

File: esp8266/modesp.c [hunk 2]
```diff
@@ -35,6 +35,492 @@
...
 #include "queue.h"
 #include "user_interface.h"
+#include "espconn.h"
+
+#include "utils.h"
+
+#define IPV4ADDR_BUFSIZE    4
+
+// Helpers from cc3200/mods/modnetwork.c
+STATIC void esp_espconn_parse_ipv4_addr(mp_obj_t addr_in, uint8_t *out_ip) {
```
Comment (dpgeorge) [inline esp8266/modesp.c]: We should try our best not to duplicate code across different ports, and this code already appears in cc3200 and stmhal separately.  Is your code an exact copy from cc3200?  If so we should factor it into a separate helper file common to all ports.

File: esp8266/modesp.c [hunk 3]
```diff
@@ -35,6 +35,492 @@
...
+    mp_obj_t tuple[2] = {
+        tuple[0] = esp_espconn_format_ipv4_addr(ip),
+        tuple[1] = mp_obj_new_int(port),
+    };
+    return mp_obj_new_tuple(2, tuple);
+}
+
+STATIC const mp_obj_type_t esp_espconn_type;
+
+typedef struct _esp_espconn_obj_t {
```
Comment (dpgeorge) [inline esp8266/modesp.c]: Since there's a lot of code for the `espconn` object I would put it in a separate source file; eg modespconn.c

File: esp8266/modesp.c [hunk 4]
```diff
@@ -116,6 +592,8 @@ STATIC const mp_map_elem_t esp_module_globals_table[] = {
     { MP_OBJ_NEW_QSTR(MP_QSTR_disconnect), (mp_obj_t)&esp_disconnect_obj },
     { MP_OBJ_NEW_QSTR(MP_QSTR_scan), (mp_obj_t)&esp_scan_obj },
     { MP_OBJ_NEW_QSTR(MP_QSTR_status), (mp_obj_t)&esp_status_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_espconn),      (mp_obj_t)&esp_espconn_type },
```
Comment (dpgeorge) [inline esp8266/modesp.c]: Unnecessary whitespace.

File: esp8266/modesp.c [hunk 5]
```diff
@@ -35,6 +35,492 @@
...
+    .locals_dict = (mp_obj_t)&esp_espconn_locals_dict,
+};
+
+typedef struct _esp_getaddrinfo_cb_struct_t {
+    mp_obj_t lambda;
+    mp_uint_t port;
+} esp_getaddrinfo_cb_struct_t;
+
+STATIC void esp_getaddrinfo_cb(const char *name, ip_addr_t *ipaddr,
+    void *arg) {
```
Comment (dpgeorge) [inline esp8266/modesp.c]: Why break the line here?  It's not that long :)

File: esp8266/modesp.c [hunk 6]
```diff
@@ -35,6 +35,492 @@
...
+    mp_obj_t lambda_in) {
+    mp_uint_t hlen;
+    const char *host = mp_obj_str_get_data(host_in, &hlen);
+    ip_addr_t ipaddr;
+
+    // Altough esp_espconn_gethostname specifies esp_espconn * as its first argument,
+    // it does not look like it actually uses it for anything so we are
+    // free to pass whatever we want
+    esp_getaddrinfo_cb_struct_t *arg =
+        gc_alloc(sizeof(esp_getaddrinfo_cb_struct_t), false);
```
Comment (dpgeorge) [inline esp8266/modesp.c]: Please use m_new_* function.

File: esp8266/modesp.c [hunk 7]
```diff
@@ -35,6 +35,492 @@
...
+    arg->port = mp_obj_get_int(port_in);
+    arg->lambda = lambda_in;
+    err_t ret = espconn_gethostbyname((void *)arg, host, &ipaddr,
+        esp_getaddrinfo_cb);
+
+    if (ret == ESPCONN_OK)
```
Comment (dpgeorge) [inline esp8266/modesp.c]: Curly braces.

File: esp8266/modesp.c [hunk 8]
```diff
@@ -35,6 +35,492 @@
...
+        tuple->items[0] = MP_OBJ_NEW_SMALL_INT(0);
+        tuple->items[1] = MP_OBJ_NEW_SMALL_INT(0);
+        tuple->items[2] = MP_OBJ_NEW_SMALL_INT(0);
+        tuple->items[3] = MP_OBJ_NEW_QSTR(MP_QSTR_);
+        tuple->items[4] = esp_espconn_format_inet_addr(ip, vals->port);
+        call_function_1_protected(vals->lambda, tuple);
+    } else {
+        call_function_1_protected(vals->lambda, mp_const_none);
+    }
+    gc_free(vals);
```
Comment (dpgeorge) [inline esp8266/modesp.c]: Please use m_del_* function.

File: esp8266/modesp.c (hunk 6 above)
Comment (dpgeorge) [inline esp8266/modesp.c]: I think this is a dangerous use of malloc: we don't know where the pointer to arg is stored when you pass it to espconn_gethostbyname, so we don't know if the GC will scan it or not.  Best probably to just use some static bss memory for this and just tell users that getaddrinfo can't be called again without the first one returning.

File: esp8266/modesp.c [hunk 9]
```diff
@@ -35,6 +35,492 @@
...
+}
+
+// constructor esp_espconn(family=AF_INET, type=SOCK_STREAM, proto=IPPROTO_TCP, fileno=None)
+// Arguments ignored as we do not support UDP (yet)
+STATIC mp_obj_t esp_espconn_make_new(mp_obj_t type_in, mp_uint_t n_args,
+    mp_uint_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 0, 4, false);
+
+    esp_espconn_obj_t *s = esp_espconn_make_new_base();
+    s->espconn = gc_alloc(sizeof(struct espconn), false);
```
Comment (dpgeorge) [inline esp8266/modesp.c]: Please use m_new_obj.

File: esp8266/modesp.c [hunk 10]
```diff
@@ -35,6 +35,492 @@
...
+// method socket.__del__()
+STATIC mp_obj_t esp_espconn___del__(mp_obj_t self_in) {
+    esp_espconn_obj_t *s = self_in;
+
+    esp_espconn_close(self_in);
+
+    if (s->server) {
+        espconn_delete(s->espconn);
+    } else {
+        gc_free(s->espconn->proto.tcp);
```
Comment (dpgeorge) [inline esp8266/modesp.c]: Please use m_del_obj.

Comment (dpgeorge) [inline esp8266/modesp.c]: Actually, you shouldn't free this memory explicitly.  If the GC has already marked the espconn object for reclamation then it already marked these buffers and freeing them explicitly may lead to undefined behaviour.

File: esp8266/modesp.c [hunk 11]
```diff
@@ -35,6 +35,492 @@
...
+    s->espconn->proto.tcp = gc_alloc(sizeof(esp_tcp), false);
+
+    return s;
+}
+
+// method espconn.close()
+STATIC mp_obj_t esp_espconn_close(mp_obj_t self_in) {
+    esp_espconn_obj_t *s = self_in;
+
+    if (esp_espconn_listening == s)
```
Comment (dpgeorge) [inline esp8266/modesp.c]: Curly braces.

File: esp8266/modesp.c (hunk 2 above)
Comment (danicampora) [inline esp8266/modesp.c]: +1 on putting it in a common file.

Comment (pfalcon) [inline esp8266/modesp.c]: +1 I understand such refactoring is a bit "hard" to do for contributors (vs maintainers), but duplicating code is the worst choice. If not complete factoring out those funcs from cc3200 code, then at least linking to them (they may be part of bigger module, so that may be hard). @atalax , so let us know if you feel like giving that a try or if we should do it first. If you try, the good location for those functions is lib/ .

File: esp8266/modesp.c [hunk 12]
```diff
@@ -35,6 +35,492 @@
...
+    mp_uint_t recvbuf_len;
+
+    bool server;
+
+    mp_obj_list_t *connlist;
+} esp_espconn_obj_t;
+
+// Due to the onconnect callback not being able to recognize the parent esp_espconn,
+// we can have only one esp_espconn listening at a time
+// This should be solvable by some PIC hacking
```
Comment (pfalcon) [inline esp8266/modesp.c]: What do you mean by "PIC hacking"?

File: esp8266/qstrdefsport.h [hunk 13]
```diff
@@ -41,11 +41,30 @@ Q(udelay)
...
 
+// esp module
 Q(esp)
+Q(espconn)
 Q(connect)
 Q(disconnect)
 Q(scan)
 Q(status)
+Q(getaddrinfo)
+Q(type)
```
Comment (pfalcon) [inline esp8266/qstrdefsport.h]: Why is this needed? I don't see QSTR_type referenced in the patch.

**Comment by pfalcon** [comment]
This goes in right direction. But when I said it's better not to put this functionality into "usocket" module, I didn't mean that we should avoid any standard terminology. Having everything in "esp" module should give enough hint that it won't work like standard socket, but well, with that note, it's rather be still called "socket", not "espconn" - there's enough similarity after all to warrant that.

getaddrinfo() then should be a module function, not a socket object method - unless there's good reason for it to be so, then please elaborate.

P.S. You code is good, let's spend more time in the discussion ;-). E.g., it would be nice to see example of code.

File: esp8266/modesp.c (hunk 12 above)
Comment (atalax) [inline esp8266/modesp.c]: Well, the solution could be:
1) Build the function as [position independent](http://en.wikipedia.org/wiki/Position-independent_code)
2) For every listening espconn, copy the function + some metadata to a new place in memory
3) Then, as the function gets called, find the metadata (we can calculate offset from start of function using [this](http://stackoverflow.com/questions/2154852/get-a-pointer-to-the-current-function-in-c-gcc) and get current pc using asm()) and dispatch to the appropriate callback using them
Thinking about it now, it would be way simpler to just use bare asm function though...

**Comment by dpgeorge** [comment]
> getaddrinfo() then should be a module function

It looks like it already is.

File: esp8266/modesp.c (hunk 2 above)
Comment (atalax) [inline esp8266/modesp.c]: I should be able to do it, or I am going to at least try :)

File: esp8266/modesp.c (hunk 12 above)
Comment (pfalcon) [inline esp8266/modesp.c]: I see, this is often called "trampoline". This comment mentioned that I envisioned need for them too: https://github.com/micropython/micropython/issues/992#issuecomment-70330542 (but it turned out that only *some*, less important functions lack ability to pass user data pointer).

The reason I asked is that, if you feel like doing hacks, with some focused and sustained effort it's possible to leave all this espconn mess behind, and have normal socket (or at least TCP/IP) API.

Anyway, it makes sense to do it step by step, starting with wrapping existing vendor stuff.

File: esp8266/modesp.c (hunk 2 above)
Comment (pfalcon) [inline esp8266/modesp.c]: It's not hard technically, only logistically - where should it be moved, how far you should go, how many ports will be broken and need fixing, etc. Thanks for trying it!

**Comment by pfalcon** [comment]
> It looks like it already is.

Oops, indeed it is, apparently I mixed contexts where looked at it.

**Comment by atalax** [comment]
Fixed all the miscellaneous stuff, moved espconn to a separate module and added another commit to move the common `mod_network_*` functions to `lib/netutils/netutils.c`. All the other modified ports build, but I do not have any HW to test them on.

**Comment by pfalcon** [comment]
> moved espconn to a separate module

There must be something wrong with my English this weekend ;-). Let me try again: the "esp" module (file modesp.c) was intended to be the single location for "all things adhoc esp8266 networking". It would contain functionality similar to modules "(u)socket" and "network" of other ports. The modules with these exact names are reserved in esp8266 for the good times when we will have fully compliant networking there. There's single module "esp" because generally having more modules leads to more overhead. As having "esp" module is already a workaround, let's keep it single, instead of providing 2 workaround modules. Within "esp" module, the names should be as similar as sensibly possible to names used in "usocket"/"network" module.

That's the idea I had. If you see any issues with the above, feel free to criticize/propose better ways.

Otherwise, this translates to: 1) please rename "espconn" type to "socket"; 2) please make sure it is accessible as "esp.socket"; 3) "espconn" module is not needed. Whether to have 1 (modesp.c, like was before) or 2 files (with esp.socket implementation split to a separate file) on C level is up to you (but please rename it to not start with "mod" then).

Great work on fixing other issues and refactoring existing code, thanks.

> All the other modified ports build, but I do not have any HW to test them on.

This is not needed for such mechanical refactoring. As long as Travis build is green, all's ok.

**Comment by danicampora** [comment]
> All the other modified ports build, but I do not have any HW to test them on

Thanks. I'll check on the CC3200 and I'll get back ASAP.

**Comment by atalax** [comment]
> There must be something wrong with my English this weekend ;-).

My fault, I missed your comment :). Everything should be refractored correctly now, I also switched the order of the commits, it makes more sense this way.

**Comment by dpgeorge** [comment]
> Whether to have 1 (modesp.c, like was before) or 2 files (with esp.socket implementation split to a separate file) on C level is up to you (but please rename it to not start with "mod" then).

It was actually my suggestion to put the espconn/esp.socket class in a separate file called modesp.c, hence the confusion here.  I should have called it modespsocket.c, to mirror modpybpin.c (Pin is a class within pyb module).

**Comment by dpgeorge** [comment]
It's a small shame to unnecessarily increase code size of cc3200 by adding the endian argument for the netutils_* functions (when cc3200 is always little endian for these calls).  But that's something which can be optimised later.

**Comment by dpgeorge** [comment]
@pfalcon I'm happy with this.  We can keep the esp.socket class in the modesp.c file for now, and pull it out later if the file gets too big.

**Comment by pfalcon** [comment]
Yes, looks good to me too.

**Comment by danicampora** [comment]
It's OK for me too.

**Comment by dpgeorge** [comment]
Rebased, tweaked and merged in 04ee5983feadc6945e324eb54cd9e7f8c1d6aa8c and 9e00ac89d5062414b618b3e961acda04ddc8a30d.

Thanks heaps @atalax!

Now that esp8266 has non-trivial features we need some test scripts (which can also serve as examples).

**Comment by danicampora** [comment]
> Now that esp8266 has non-trivial features we need some test scripts (which can also serve as examples).

This is also my next step with the CC3200

**Comment by dpgeorge** [comment]
> This is also my next step with the CC3200

It's not easy, eg you need to somehow configure your AP without storing passwords in the tests :)

We can have common network/socket tests, but most won't work with the ESP due to the different/non-standard socket API.

A proper CI testing scheme would be idea, with all supported boards connected to test harnesses on a server... but that's not going to happen anytime soon :)

**Comment by pfalcon** [comment]
> Now that esp8266 has non-trivial features we need some test scripts (which can also serve as examples).

I looked at the code recently (after haven't looking for 2 months), and saw that non-friendly REPL is still sync version, which won't work right (even though I definitely started refactoring to even-driven with it, as it's simpler).

**Comment by pfalcon** [comment]
> A proper CI testing scheme

Let's hope that @danicampora will drive that area, as he's a vendor now ;-).

**Comment by danicampora** [comment]
> It's not easy, eg you need to somehow configure your AP without storing passwords in the tests :)

This part can be in a config file in the test machine itself. A template/example can be in the repo.

> A proper CI testing scheme would be idea, with all supported boards connected to test harnesses on a server... but that's not going to happen anytime soon :)

Well, that's what I'm going to do. Right after the KS I'll start putting up a test machine built with a raspberry PI 2 running Jenkins. Where I work, we have similar setups, and they work well. I will use Jenkins because that's what I'm familiar with, unless somebody else has a better suggestion ;-)
